### PR TITLE
DDPB-4293: Update Lay CSV upload and reorganise app around pre-registration rather than casrec

### PR DESCRIPTION
**Pull request by acsauk**
## Purpose
Given Casrec's final days are (finally) upon us we need to update the CSV upload process to support a CSV in a new format and scrub any signs that casrec existed 🔪 . Given the table the CSV is loaded to supports the Lay registration process we've gone with `PreRegistration` as the entity name and have updated all refs of casrec to either pre-registration or Sirius based on the context.

I've left a couple of references to casrec in regards to org CSV uploads which we will tackle separately.

Fixes DDPB-4293

## Approach
This was a lot of find and replace followed by test runs to see what had been missed. Rinse and repeat.

## Checklist
- [x] I have performed a self-review of my own code
- [ ] I have updated documentation (Confluence/ADR/tech debt doc) where relevant
- [x] I have added tests to prove my work
- [ ] The product team have approved these changes

### CODE REVIEW

File: api/src/FixtureFactory/PreRegistrationFactory.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        $dto = (new LayDeputyshipDto())
+            ->setCaseNumber($data['caseNumber'] ?: $caseNumber)
+            ->setClientSurname($data['clientLastName'] ?: 'Smith')
+            ->setDeputyUid($deputyNumber)
+            ->setDeputyAddress1($data['deputyPostCode'] ?: 'Victoria Park')
+            ->setDeputyAddress2($data['deputyPostCode'] ?: 'Fakeville')
+            ->setDeputyAddress3($data['deputyPostCode'] ?: 'Pretendham')
+            ->setDeputyAddress4($data['deputyPostCode'] ?: null)
+            ->setDeputyAddress5($data['deputyPostCode'] ?: null)
+            ->setDeputyPostcode($data['deputyPostCode'] ?: 'SW1')
```
Comment (acsauk) [inline api/src/FixtureFactory/PreRegistrationFactory.php:35]: Update to proper fields

File: api/src/Migrations/Version262.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        $this->addSql('CREATE TABLE pre_registration (id SERIAL NOT NULL, client_case_number VARCHAR(20) NOT NULL, client_lastname VARCHAR(50) NOT NULL, deputy_uid VARCHAR(100) NOT NULL, deputy_lastname VARCHAR(100) DEFAULT NULL, deputy_address_1 VARCHAR(255) DEFAULT NULL, deputy_address_2 VARCHAR(255) DEFAULT NULL, deputy_address_3 VARCHAR(255) DEFAULT NULL, deputy_address_4 VARCHAR(255) DEFAULT NULL, deputy_address_5 VARCHAR(255) DEFAULT NULL, deputy_postcode VARCHAR(10) DEFAULT NULL, type_of_report VARCHAR(10) DEFAULT NULL, ndr BOOLEAN DEFAULT NULL, uploaded_at TIMESTAMP(0) WITHOUT TIME ZONE DEFAULT NULL, updated_at TIMESTAMP(0) WITHOUT TIME ZONE DEFAULT NULL, order_date TIMESTAMP(0) WITHOUT TIME ZONE DEFAULT NULL, order_type VARCHAR(255) DEFAULT NULL, is_co_deputy BOOLEAN DEFAULT NULL, PRIMARY KEY(id))');
+        $this->addSql('CREATE INDEX updated_at_index ON pre_registration (updated_at)');
+        $this->addSql('DROP TABLE casrec');
+        $this->addSql('ALTER TABLE checklist ADD deputy_full_name_accurate_in_sirius VARCHAR(3) DEFAULT NULL');
+        $this->addSql('ALTER TABLE checklist ADD bond_order_match_sirius VARCHAR(3) DEFAULT NULL');
```
Comment (acsauk) [inline api/src/Migrations/Version262.php:24]: Change to `RENAME COLUMN`

File: api/src/Service/PreRegistrationVerificationService.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        $detailsToMatchOn = [
+            'caseNumber' => $caseNumber,
+            'clientLastname' => $clientSurname,
+            'deputySurname' => $deputySurname,
+        ];
+
+        /** @var PreRegistration[] $crMatches */
+        $allDetailsMatches = $this->preRegistrationRepository->findBy($detailsToMatchOn);
+
+        $this->lastMatchedPreRegistrationUsers = $this->applyPostcodeFilter($allDetailsMatches, $deputyPostcode);
```
Comment (acsauk) [inline api/src/Service/PreRegistrationVerificationService.php:47]: Apply normalisation here

File: api/src/Service/ReportService.php [hunk 4]
```diff
@@ -330,10 +320,10 @@ private function createNextYearReport(ReportInterface $oldReport)
      *
      * @throws Exception
      */
-    public function getReportTypeBasedOnCasrec(Client $client)
+    public function getReportTypeBasedOnSirius(Client $client)
     {
-        $casRec = $this->casRecRepository->findOneBy(['caseNumber' => $client->getCaseNumber()]);
-        if ($casRec instanceof CasRec) {
+        $casRec = $this->preRegistrationRepository->findOneBy(['caseNumber' => $client->getCaseNumber()]);
+        if ($casRec instanceof PreRegistration) {
```
Comment (acsauk) [inline api/src/Service/ReportService.php:326]: Update to `$preRegistration`

File: api/src/v2/Registration/Uploader/LayDeputyshipUploader.php [hunk 5]
```diff
@@ -68,55 +62,51 @@ public function upload(LayDeputyshipDtoCollection $collection): array
...
-            break;
-        }
-
         return [
             'added' => $added,
             'errors' => $errors,
             'report-update-count' => count($this->reportsUpdated),
             'cases-with-updated-reports' => $this->reportsUpdated,
-            'source' => $source,
+            'source' => 'sirius',
```
Comment (acsauk) [inline api/src/v2/Registration/Uploader/LayDeputyshipUploader.php:74]: Drop as not needed

File: api/tests/Behat/features-v2/report-submission/report-submission.feature [hunk 6]
```diff
@@ -1,4 +1,4 @@
-@report-submissions @v2_admin @v2
+@report-submissions @v2_admin @v2 @acs
```
Comment (acsauk) [inline api/tests/Behat/features-v2/report-submission/report-submission.feature:1]: Drop debug tag

File: client/src/Controller/ClientController.php [hunk 7]
```diff
@@ -151,11 +137,11 @@ public function addAction(Request $request, Redirector $redirector, TranslatorIn
...
             try {
-                // validate against casRec
-                $this->restClient->apiCall('post', 'casrec/verify', $client, 'array', []);
+                // validate against pre-registration data
+                $this->preRegistrationApi->verify($client);
 
-                // $method is set above to either post or put
-                $response = $this->restClient->$method('client/upsert', $form->getData());
+                $response = 'post' === $method ?
+                    $this->clientApi->create($form->getData()) : $this->clientApi->update($client, $form->getData(), AuditEvents::TRIGGER_DEPUTY_USER_ADD_CLIENT_IN_REGISTRATION);
```
Comment (acsauk) [inline client/src/Controller/ClientController.php:144]: Break out ternaray

File: client/src/Entity/Report/Checklist.php [hunk 8]
```diff
@@ -653,19 +653,19 @@ public function setBondAdequate($bondAdequate)
...
-    public function getBondOrderMatchCasrec()
+    public function getBondOrderMatchSirius()
     {
-        return $this->bondOrderMatchCasrec;
+        return $this->bondOrderMatchSirius;
     }
 
     /**
-     * @param string $bondOrderMatchCasrec
+     * @param string $bondOrderMatchSirius
```
Comment (acsauk) [inline client/src/Entity/Report/Checklist.php:662]: Double check twig for refs to prop names

File: client/src/Service/Audit/AuditEvents.php [hunk 9]
```diff
@@ -28,6 +29,7 @@ final class AuditEvents
     const TRIGGER_ADMIN_USER_EDIT = 'ADMIN_USER_EDIT';
     const TRIGGER_ADMIN_BUTTON = 'ADMIN_BUTTON';
     const TRIGGER_CSV_UPLOAD = 'CSV_UPLOAD';
+    const TRIGGER_DEPUTY_USER_ADD_CLIENT_IN_REGISTRATION = 'DEPUTY_USER_ADD_CLIENT_IN_REGISTRATION';
```
Comment (acsauk) [inline client/src/Service/Audit/AuditEvents.php:32]: Do we need this?

File: api/src/Controller/Report/ReportController.php [hunk 10]
```diff
@@ -778,7 +778,7 @@ private function populateChecklistEntity($checklist, $checklistData)
             'accounts_balance' => 'setAccountsBalance',
             'assets_declared_and_managed' => 'setAssetsDeclaredAndManaged',
             'bond_adequate' => 'setBondAdequate',
-            'bond_order_match_casrec' => 'setBondOrderMatchCasrec',
+            'bond_order_match_sirius' => 'setbondOrderMatchSirius',
```
Comment (Gugandeep) [inline api/src/Controller/Report/ReportController.php:781]: setbondOrderMatchSirius -> setBondOrderMatchCasrec

**Pull request merged**